src/Formats/IonWriter.cpp

```diff
@@ -0,0 +1,306 @@
+#include "IonWriter.h"
+
+#if USE_ION
+#include <decNumber/decContext.h>
+#include <Common/formatIPv6.h>
+
+namespace DB
+{
+
+IonWriter::IonWriter(WriteBuffer & out_, bool output_as_binary_) : out(out_), ion_buffer(out.available())
+{
+    // init output options
+    ION_WRITER_OPTIONS ion_options;
+    memset(&ion_options, 0, sizeof(ion_options));
+    ion_options.output_as_binary = output_as_binary_; // todo: Fix switch of writer type from format settings
+    if (!output_as_binary_)
+        ion_options.pretty_print = TRUE;
+
+    // Callback function called when flushing from an output stream
+    // flash is called only if we are at the zero nesting level, that is, not in a list or structure,
+    // which means one full ion message will be saved to the internal buffer,
+    // and then it will be transferred to WriteBuffer
+    ion_writer_callback = [this](_ion_user_stream * stream) -> iERR
+    {
+        // if limit is init it is call from flush
+        if (stream->limit == ion_buffer.data() + ion_buffer.size())
+        {
+            // Save current total written bytes
+            if ((stream->curr >= ion_buffer.data()) && (stream->curr <= ion_buffer.data() + ion_buffer.size()))
+                total_written_bytes = stream->curr - ion_buffer.data();
+
+            // Checking whether the buffer has overflowed; if so, expand and update the stream pointers
+            if (stream->curr == stream->limit)
+            {
+                // todo: Should we put a border on top of this?
+                ion_buffer.resize(ion_buffer.size() << 1);
+                stream->limit = ion_buffer.data() + ion_buffer.size();
+                stream->curr = ion_buffer.data() + total_written_bytes;
+            }
+        }
+        else
+        {
+            // else it is first call
+            total_written_bytes = 0;
+            stream->limit = ion_buffer.data() + ion_buffer.size();
+            stream->curr = ion_buffer.data();
+        }
+        return IERR_OK;
+    };
+
+    // Init stream and writer
+    ION_STREAM * stream = nullptr;
+    hWRITER writer = nullptr;
+    auto err = ion_stream_open_handler_out(&ionWriteCallback, &ion_writer_callback, &stream);
+    if (err == IERR_OK)
+    {
+        err = ion_writer_open(&writer, stream, &ion_options);
+        if (err == IERR_OK)
+        {
+            ion_stream = stream;
+            ion_writer = writer;
+        }
+        else
+        {
+            ion_stream_close(stream);
+            // todo: Should we be doing anything else here?
+        }
+    }
```

Comment:
If `err != IERR_OK` we should use `ion_error_to_str` function and throw an exception with this message. We can add new error code in `Common/ErrorCodes.cpp` for it and name it `ION_FORMAT_WRITER_ERROR`, so we will use it like:
```
throw Exception(ErrorCodes::ION_FORMAT_WRITER_ERROR, "Cannot create ION format writer, error: {}", ion_error_to_str(...))
```